File: web/src/api/tracks.ts
import type { PaginatedResult, Track, TrackFilters } from "../types";
import client from "./client";

function buildParams(f: TrackFilters): Record<string, string | string[]> {
  const params: Record<string, string | string[]> = {};
  if (f.search) params.search = f.search;
  if (f.genres?.length) params.genre = f.genres;
  if (f.year_min != null) params.year_min = String(f.year_min);
  if (f.year_max != null) params.year_max = String(f.year_max);
  if (f.popularity_min != null) params.popularity_min = String(f.popularity_min);
  if (f.popularity_max != null) params.popularity_max = String(f.popularity_max);
  if (f.duration_min != null) params.duration_min = String(f.duration_min);
  if (f.duration_max != null) params.duration_max = String(f.duration_max);
  if (f.explicit != null) params.explicit = String(f.explicit);
  if (f.playlist) params.playlist = f.playlist;
  if (f.saved_at_min) params.saved_at_min = f.saved_at_min;
  if (f.saved_at_max) params.saved_at_max = f.saved_at_max;
  if (f.artist_popularity_min != null) params.artist_popularity_min = String(f.artist_popularity_min);
  if (f.artist_popularity_max != null) params.artist_popularity_max = String(f.artist_popularity_max);
  if (f.artist_followers_min != null) params.artist_followers_min = String(f.artist_followers_min);
  if (f.artist_followers_max != null) params.artist_followers_max = String(f.artist_followers_max);
  if (f.tempo_min != null) params.tempo_min = String(f.tempo_min);
  if (f.tempo_max != null) params.tempo_max = String(f.tempo_max);
  if (f.energy_min != null) params.energy_min = String(f.energy_min);
  if (f.energy_max != null) params.energy_max = String(f.energy_max);
  if (f.danceability_min != null) params.danceability_min = String(f.danceability_min);
  if (f.danceability_max != null) params.danceability_max = String(f.danceability_max);
  if (f.valence_min != null) params.valence_min = String(f.valence_min);
  if (f.valence_max != null) params.valence_max = String(f.valence_max);
  if (f.acousticness_min != null) params.acousticness_min = String(f.acousticness_min);
  if (f.acousticness_max != null) params.acousticness_max = String(f.acousticness_max);
  if (f.instrumentalness_min != null) params.instrumentalness_min = String(f.instrumentalness_min);
  if (f.instrumentalness_max != null) params.instrumentalness_max = String(f.instrumentalness_max);
  if (f.liveness_min != null) params.liveness_min = String(f.liveness_min);
  if (f.liveness_max != null) params.liveness_max = String(f.liveness_max);
  if (f.speechiness_min != null) params.speechiness_min = String(f.speechiness_min);
  if (f.speechiness_max != null) params.speechiness_max = String(f.speechiness_max);
  if (f.loudness_min != null) params.loudness_min = String(f.loudness_min);
  if (f.loudness_max != null) params.loudness_max = String(f.loudness_max);
  if (f.keys?.length) params.key = f.keys.map(String);
  if (f.mode != null) params.mode = String(f.mode);
  if (f.time_signatures?.length) params.time_signature = f.time_signatures.map(String);
  if (f.page) params.page = String(f.page);
  if (f.page_size) params.page_size = String(f.page_size);
  if (f.sort_by) params.sort_by = f.sort_by;
  if (f.sort_dir) params.sort_dir = f.sort_dir;
  return params;
}

export async function getTracks(filters: TrackFilters = {}): Promise<PaginatedResult<Track>> {
  const res = await client.get<PaginatedResult<Track>>("/tracks", { params: buildParams(filters) });
  return res.data;
}

File: server/internal/handler/tracks.go
package handler

import (
	"context"
	"encoding/json"
	"net/http"
	"net/url"
	"strconv"
	"time"

	"github.com/mcopland/spotifind/internal/middleware"
	"github.com/mcopland/spotifind/internal/models"
)

// TrackQuerier is satisfied by repository.TrackRepo.
type TrackQuerier interface {
	ListForUser(ctx context.Context, userID int64, f models.TrackFilters) (*models.PaginatedResult[models.Track], error)
	GetStatsForUser(ctx context.Context, userID int64) (*models.TrackStats, error)
}

type TrackHandler struct {
	trackRepo TrackQuerier
}

func NewTrackHandler(repo TrackQuerier) *TrackHandler {
	return &TrackHandler{trackRepo: repo}
}

func (h *TrackHandler) List(w http.ResponseWriter, r *http.Request) {
	userID, ok := middleware.GetUserID(r.Context())
	if !ok {
		http.Error(w, "unauthorized", http.StatusUnauthorized)
		return
	}

	f := parseTrackFilters(r)
	result, err := h.trackRepo.ListForUser(r.Context(), userID, f)
	if err != nil {
		http.Error(w, "failed to list tracks", http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(result)
}

func (h *TrackHandler) Stats(w http.ResponseWriter, r *http.Request) {
	userID, ok := middleware.GetUserID(r.Context())
	if !ok {
		http.Error(w, "unauthorized", http.StatusUnauthorized)
		return
	}

	stats, err := h.trackRepo.GetStatsForUser(r.Context(), userID)
	if err != nil {
		http.Error(w, "failed to get track stats", http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(stats)
}

func parseTrackFilters(r *http.Request) models.TrackFilters {
	q := r.URL.Query()
	f := models.TrackFilters{
		Search:     q.Get("search"),
		PlaylistID: q.Get("playlist"),
		SortBy:     q.Get("sort_by"),
		SortDir:    q.Get("sort_dir"),
	}
	if v := q.Get("artist_id"); v != "" {
		f.ArtistSpotifyID = &v
	}

	if genres := q["genre"]; len(genres) > 0 {
		f.Genres = genres
	}

	parseInt(q, "year_min", &f.YearMin)
	parseInt(q, "year_max", &f.YearMax)
	parseInt(q, "popularity_min", &f.PopularityMin)
	parseInt(q, "popularity_max", &f.PopularityMax)
	parseInt(q, "duration_min", &f.DurationMin)
	parseInt(q, "duration_max", &f.DurationMax)
	parseInt(q, "artist_popularity_min", &f.ArtistPopularityMin)
	parseInt(q, "artist_popularity_max", &f.ArtistPopularityMax)
	parseInt(q, "artist_followers_min", &f.ArtistFollowersMin)
	parseInt(q, "artist_followers_max", &f.ArtistFollowersMax)

	parseFloat(q, "tempo_min", &f.TempoMin)
	parseFloat(q, "tempo_max", &f.TempoMax)
	parseFloat(q, "energy_min", &f.EnergyMin)
	parseFloat(q, "energy_max", &f.EnergyMax)
	parseFloat(q, "danceability_min", &f.DanceabilityMin)
	parseFloat(q, "danceability_max", &f.DanceabilityMax)
	parseFloat(q, "valence_min", &f.ValenceMin)
	parseFloat(q, "valence_max", &f.ValenceMax)
	parseFloat(q, "acousticness_min", &f.AcousticnessMin)
	parseFloat(q, "acousticness_max", &f.AcousticnessMax)
	parseFloat(q, "instrumentalness_min", &f.InstrumentalnessMin)
	parseFloat(q, "instrumentalness_max", &f.InstrumentalnessMax)
	parseFloat(q, "liveness_min", &f.LivenessMin)
	parseFloat(q, "liveness_max", &f.LivenessMax)
	parseFloat(q, "speechiness_min", &f.SpeechinessMin)
	parseFloat(q, "speechiness_max", &f.SpeechinessMax)
	parseFloat(q, "loudness_min", &f.LoudnessMin)
	parseFloat(q, "loudness_max", &f.LoudnessMax)

	parseTime(q, "saved_at_min", &f.SavedAtMin)
	parseTime(q, "saved_at_max", &f.SavedAtMax)

	if v := q.Get("explicit"); v != "" {
		b := v == "true"
		f.Explicit = &b
	}
	if v := q.Get("mode"); v != "" {
		if n, err := strconv.Atoi(v); err == nil {
			f.Mode = &n
		}
	}
	if keys := q["key"]; len(keys) > 0 {
		f.Keys = parseIntList(keys)
	}
	if sigs := q["time_signature"]; len(sigs) > 0 {
		f.TimeSignatures = parseIntList(sigs)
	}

	if v := q.Get("page"); v != "" {
		if n, err := strconv.Atoi(v); err == nil {
			f.Page = n
		}
	}
	if v := q.Get("page_size"); v != "" {
		if n, err := strconv.Atoi(v); err == nil {
			f.PageSize = n
		}
	}

	return f
}

func parseInt(q url.Values, key string, dst **int) {
	v := q.Get(key)
	if v == "" {
		return
	}
	n, err := strconv.Atoi(v)
	if err != nil {
		return
	}
	*dst = &n
}

func parseFloat(q url.Values, key string, dst **float64) {
	v := q.Get(key)
	if v == "" {
		return
	}
	n, err := strconv.ParseFloat(v, 64)
	if err != nil {
		return
	}
	*dst = &n
}

func parseTime(q url.Values, key string, dst **time.Time) {
	v := q.Get(key)
	if v == "" {
		return
	}
	t, err := time.Parse(time.RFC3339, v)
	if err != nil {
		return
	}
	*dst = &t
}

func parseIntList(values []string) []int {
	out := make([]int, 0, len(values))
	for _, v := range values {
		n, err := strconv.Atoi(v)
		if err != nil {
			continue
		}
		out = append(out, n)
	}
	return out
}
